docs/SPIRVVersionsAndExtensionsHandling.rst

```diff
@@ -0,0 +1,247 @@
+valid SPIR-V according to the core spec, without uses of any extensions if such
+SPIR-V was requested by user.
+
+For example, without such infrastructure it is impossible to disable use of
+``SPV_KHR_no_integer_wrap_decoration`` - it will be always generated if
+corresponding LLVM IR counterparts are encountered in input module.
+
+It is worth mentioning that SPIR-V versions and extensions the handling of
+SPIR-V versions and extension is mostly important for the SPIR-V generation
+step. On the consumer side it is the responsibility of the consumer to analyze
+the incoming SPIR-V file and reject it if it contains something that is not
+supported by the consumer.
+
+However, translator simplifies this step for downstream users by checking
+version and extensions in SPIR-V module during ``readSpirv``/``readSpirvModule``
+phases.
+
+SPIR-V Versions
+===============
+
+SPIR-V Generation step
+----------------------
+
+By default translator selects version of generated SPIR-V file based on features
+used in this file. For example, if it contains ``dereferencable`` LLVM IR
+attribute, ``MaxByteOffset`` decoration will be generated and resulting SPIR-V
+version will be raised to 1.1.
+
+.. note::
+   There is no documentation about which exact features from newest
+   SPIR-V spec versions will be used by the translator. If you are interested
+   when or why a particular SPIR-V instruction is generated, please check this
+   in the source code. Consider this as an implementation detail and if you
+   disagree with something, you can always open an issue or submit pull request
+   - contributions are welcome!
+
+There is one option to control the behavior of the translator with respect to
+the version of the SPIR-V file which is being generated/consumed.
+
+* ``--spirv-max-version=`` - instructs the translator to generate SPIR-V file
+  corresponding to any spec version which is less than or equal to the
+  specified one. Behavior of the translator is the same as by default with only
+  one exception: resulting SPIR-V version cannot be raised higher than
+  specified by this option.
+
+Allowed values are ``1.0`` and ``1.1``.
+
+.. warning::
+   These two options are mutually exclusive and cannot be specified at the
+   same time.
+
+If the translator encounters something that cannot be represented by set of
+allowed SPIR-V versions (which might contain only one version), it does one of
+the following things:
+
+* ignores LLVM IR entity in the input file.
+
+  For example, ``dereferencable`` LLVM IR attribute can be ignored if it is not
+  allowed to generate SPIR-V 1.1 and higher.
+
+* tries to represent LLVM IR entity with allowed instructions.
+
+  For example, ``OpPtrEqual`` can be used if SPIR-V 1.4 is not allowed and can
+  be emulated via ``OpConvertPtrToU`` + ``OpIEqual`` sequence.
+
+* emits error if LLVM IR entity cannot be ignored and cannot be emulated using
+  available instructions.
+
+  **TODO**: Example
```

Comment:
Are you planning to fill in the TODO before merging this PR?
(and there are actually other TODOs in this PR too)

Reply:
Here is the list of TODOs/FIXMEs in my PR with my plans around them:

* examples in documentation - will fix them in this PR
* LLVMSPIRVOpts.h:53 - not going to fix in this PR. Should be fixed when we update translator to support SPIR-V 1.2 and higher
* SPIRVValue - not going to fix in this PR. Should be fixed when we update the translator to support SPIR-V 1.4
* the rest are related to IntelFPGA-specific extensions and I also don't want to increase this PR with tests and translator improvements - let's say this PR introduces basic infrastructure and fixes LIT tests, but the rest of improvements that can be built on top of this infrastructure will go as separate PRs